dataset-processor-python/data_loader.py

```diff
@@ -1,55 +1,298 @@
+        """
+        Returns the column data from the dataset
+        """
+        column_data = []
+        for row in self.data:
+            column_data.append(row.get(column, ""))
+        return column_data
+
+    @staticmethod
+    def get_color(index):
+        colors = [
+            "#A6CEE3",
+            "#1F78B4",
+            "#B2DF8A",
+            "#33A02C",
+            "#FB9A99",
+            "#E31A1C",
+            "#FDBF6F",
+            "#FF7F00",
+            "#CAB2D6",
+        ]
+        if index >= len(colors):
+            index = index % len(colors)
+        return colors[index]
+
+    def write_discrete_feature_options(self, data_values):
+        """
+        Keys in options are the unique values in the column, and values are the color and name of the feature
+        Returns the options for a discrete feature
+
+        Options example:
+        "options":{
+            "7": {
+                "color": "#FF96FF",
+                "name": "Actin filaments",
+                "key": "Alpha-actinin-1"
+            },
+            "16": {
+                "color": "#FFB1FF",
+                "name": "Actin filaments",
+                "key": "Beta-actin"
+        }
+        """
+        keys = numpy.unique(data_values)
+        options = {}
+        for index, key in enumerate(keys):
+            # "key" is optional, if "name" is not unique, use "key" to distinguish between the options
+            options[key] = {"color": self.get_color(index), "name": "", "key": ""}
+        return options
+
+    def compile_feature_defs(self):
+        description = ""
+        tooltip = ""
+        for key in self.feature_def_keys:
+            discrete = self.discrete_features_dict[key]
+            unit = self.get_unit(key)
+            stripped_key = key.replace(f"({unit})", "").strip()
+            self.features_data_order.append(stripped_key)
+            display_name = DatasetWriter.format_display_name(
+                stripped_key.replace("-", " ")
+            )
+            feature_def = {
+                "key": stripped_key,
+                "displayName": display_name,
+                "unit": unit,
+                "description": description,
+                "tooltip": tooltip,
+                "discrete": discrete,
+            }
+            if discrete:
+                column_data = self.get_column_data(key)
+                options = self.write_discrete_feature_options(column_data)
+                feature_def["options"] = options
+            self.feature_defs_data.append(feature_def)
+
+    def compile_dataset(self):
+        fields_to_write = {
+            "title": self.inputs["title"],
+            "version": self.inputs["version"],
+            "name": self.inputs["dataset_name"],
+            "description": self.inputs["description"],
+            "featureDefsPath": "feature_defs.json",
+            "featuresDataPath": "cell_feature_analysis.json",
+            "viewerSettingsPath": "image_settings.json",
```

Comment:
Nit: These could likely be made into file constants, since they're shared with lines 287-290